### DISCUSSION TITLE
Merging OrganizeImports into Scalafix as a built-in rule

### DISCUSSION THREAD

**Discussion by liancheng** [liancheng/scalafix-organize-imports#215] [General]
Here I'd like to start a thread for merging `OrganizeImports` into Scalafix as a built-in rule and discussing possible work items needed to make it happen.

Motivations:

1. Imports organization used to be a highly demanded Scalafix feature. `OrganizeImports` filled the gap and is now bundled by Metals. Making it a built-in rule of Scalafix removes a dependency from Metals and makes it easier to evolve this rule together with the ecosystem.
2. `OrganizeImports` is now relatively stable and feature-complete (Scala 3 support is still experimental and not shipped with any release yet, though).
3. I'm personally no longer capable of maintaining this project because

   - I do not have enough domain knowledge in sbt and Scala 3 to build/fix relevant new features/issues. Recent relevant changes were all accomplished by contributors like @tgodzik and @bjaglin.
   - I no longer have enough free cycles for maintaining this project, while I believe it deserves some more love.

Cc @bjaglin @tgodzik @olafurpg @mlachkar. Please let me know whether this proposal is feasible, and if yes, what could be the next steps?

**Comment by olafurpg** [comment]
Sounds good to me 👍 thank you for starting this discussion! This makes a lot of sense especially since OrganizeImports has special handling in Metals. 

Ive been using OrganizeImports in all of my projects for a while now and it’s working great.

**Reply by tgodzik** [reply]
I think it makes sense, it's pretty stable and used in a lot of different places. Especially if it would help out with maintaining the project.

Let us know if you need help with migrating it!

**Reply by liancheng** [reply]
Thanks, @olafurpg and @tgodzik! Let me try to put up a draft PR first. I anticipate that I'll need some help on the sbt build (due to lack of domain knowledge) and docs (due to lack of cycles).

**Reply by mlachkar** [reply]
Hello! We will be happy to help you!

**Reply by bjaglin** [reply]
> I anticipate that I'll need some help on the sbt build

I think a preliminary work would be to port what was done in https://github.com/liancheng/scalafix-organize-imports/pull/179 in the scalafix build itself, so that the test structure can be merged "as-is". This is something we wanted to do anyway, I can look at that next week.

Regarding the migration phase, I believe we can keep the same rule name (`OrganizeImports`) when inlining it. We can log a warning and shadow the community one if it's still in the classpath by doing adhoc processing in https://github.com/scalacenter/scalafix/blob/a3cfa866ae173387ce9e2b771003d9f44155bd2b/scalafix-reflect/src/main/scala/scalafix/v1/RuleDecoder.scala#L45.

**Reply by liancheng** [reply]
Made a draft PR https://github.com/scalacenter/scalafix/pull/1480 with a TODO list. Let's continue the discussion there.

**Comment by tgodzik** [comment]
Could we do a last release with the recent fixes for Scala 3 before merging to main scalafix?

**Reply by liancheng** [reply]
Just release v0.6.0 to include Scala 3 support.

**Reply by liancheng** [reply]
Please note that issue #221 is left unresolved as a known issue.

**Reply by tgodzik** [reply]
Thanks!

**Comment by bjaglin** [comment]
Finally, [Scalafix 0.11.0](https://github.com/scalacenter/scalafix/releases/tag/v0.11.0) includes `OrganizeImports` as a built-in rule! Thank you so much @liancheng for all your work on getting the rule to that level of maturity :bow: 

Should we update the README (the docs was migrated in https://github.com/scalacenter/scalafix/pull/1480), transfer the issues and archive the repo?

**Reply by daddykotex** [reply]
Thanks @bjaglin for that final effort!

**Reply by liancheng** [reply]
Thrilled that this finally happened!! Thank you so much for getting this rule into Scalafix, @bjaglin!

I'm updating the README and archiving this repo soon :)
